rmw_zenoh_cpp/src/rmw_zenoh.cpp

```diff
@@ -1657,10 +1894,93 @@ rmw_send_request(
   const void * ros_request,
   int64_t * sequence_id)
 {
-  static_cast<void>(client);
-  static_cast<void>(ros_request);
-  static_cast<void>(sequence_id);
-  return RMW_RET_UNSUPPORTED;
+  RCUTILS_LOG_INFO_NAMED(
+    "rmw_zenoh_cpp", "[rmw_send_request]");
+
+  RMW_CHECK_ARGUMENT_FOR_NULL(client, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_ARGUMENT_FOR_NULL(ros_request, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_ARGUMENT_FOR_NULL(sequence_id, RMW_RET_INVALID_ARGUMENT);
+
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    client,
+    client->implementation_identifier,
+    rmw_zenoh_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+
+  RMW_CHECK_FOR_NULL_WITH_MSG(
+    client->data,
+    "client implementation pointer is null",
+    RMW_RET_INVALID_ARGUMENT);
+
+  auto * client_data = static_cast<rmw_client_data_t *>(client->data);
+
+  rmw_context_impl_s * context_impl = static_cast<rmw_context_impl_s *>(
+    client_data->context->impl);
+
+  // Serialize data
+
+  rcutils_allocator_t * allocator = &(client_data->context->options.allocator);
+
+  size_t max_data_length = (
```

Comment:
The codeblock from here 1955 is repeated in many places in the codebase. Can we refactor it out into a function that can be called wherever needed?

Reply:
I created a function with firm:
```cpp
rmw_ret_t serialize_ros_message(rcutils_allocator_t * allocator, TypeSupport * type_support, const void * type_support_impl, const void * ros_message, char * msg_bytes, size_t * data_length)
```
see details for the implementation.
<details>

```cpp
//==============================================================================
// Serialize a ROS message into a buffer.
// Typically used by rmw_send_request (client side) and rmw_send_response (server side) to
// serialize the message before sending it over the wire.
rmw_ret_t serialize_ros_message(rcutils_allocator_t * allocator, TypeSupport * type_support, const void * type_support_impl, const void * ros_message, char * msg_bytes, size_t * data_length)
{
  // Assertions
  RMW_CHECK_FOR_NULL_WITH_MSG(
    allocator, "allocator is null",
    return RMW_RET_INVALID_ARGUMENT);
  RMW_CHECK_FOR_NULL_WITH_MSG(
    type_support, "type support handle is null",
    return RMW_RET_INVALID_ARGUMENT);
  RMW_CHECK_FOR_NULL_WITH_MSG(
    type_support_impl, "type support impl is null",
    return RMW_RET_INVALID_ARGUMENT);
  RMW_CHECK_FOR_NULL_WITH_MSG(
    ros_message, "ros message is null",
    return RMW_RET_INVALID_ARGUMENT);
  RMW_CHECK_FOR_NULL_WITH_MSG(
    data_length, "data length is null",
    return RMW_RET_INVALID_ARGUMENT);

  // Serialize data.
  size_t max_data_length = type_support->get_estimated_serialized_size(
    ros_message,
    type_support_impl);

  msg_bytes = static_cast<char *>(allocator->allocate(
      max_data_length,
      allocator->state));
  if (!msg_bytes) {
    RMW_SET_ERROR_MSG("failed allocate request message bytes");
    allocator->deallocate(msg_bytes, allocator->state);
    return RMW_RET_ERROR;
  }

  // Object that manages the raw buffer
  eprosima::fastcdr::FastBuffer fastbuffer(msg_bytes, max_data_length);
  // Object that serializes the data
  eprosima::fastcdr::Cdr ser(
    fastbuffer,
    eprosima::fastcdr::Cdr::DEFAULT_ENDIAN,
    eprosima::fastcdr::Cdr::DDS_CDR);
  if (!type_support->serialize_ros_message(
      ros_message,
      ser,
      type_support_impl))
  {
    allocator->deallocate(msg_bytes, allocator->state);
    return RMW_RET_ERROR;
  }
  *data_length = ser.getSerializedDataLength();

  return RMW_RET_OK;
}
```
</details>

And I used it in the `rmw_send_request` and `rmw_send_response` functions.

I got an odd behaviour with the `msg_bytes` pointer. **After the method is returned the `msg_bytes` pointer becomes null** even when the ownership isn't from that function. 

According to the FastBuffer docstring:
```cpp
    /*!
     * @brief This constructor assigns the user's stream of bytes to the eprosima::fastcdr::FastBuffers object.
     * The user's stream will be used to serialize.
     *
     * @param buffer The user's buffer that will be used. This buffer is not deallocated in the object's destruction. Cannot be NULL.
     * @param bufferSize The length of user's buffer.
     */
    FastBuffer(
            char* const buffer,
            const size_t bufferSize);
```

So the buffer pointer shouldn't be destructed after going off scope.